flowey/flowey_lib_common/src/publish_test_results.rs

```diff
@@ -0,0 +1,251 @@
+                                            .expect("failed to read attachment dir")
+                                            .next()
+                                            .is_some())
+                            });
+                            let attachment_path = attachment_path.map(ctx, |p| {
+                                p.absolute().expect("invalid path").display().to_string()
+                            });
+
+                            use_side_effects.push(ctx.reqv(|v| {
+                                crate::gh_task_publish_artifact::Request {
+                                    step_name: format!(
+                                        "publish test results: {label} ({attachment_label})"
+                                    ),
+                                    artifact_name: format!("{label}-{idx}-{attachment_label}"),
+                                    path: attachment_path,
+                                    condition: Some(attachment_exists),
+                                    done: v,
+                                }
+                            }));
+                        }
+                    }
+                }
+                ctx.emit_side_effect_step(use_side_effects, resolve_side_effects);
+            }
+            FlowBackend::Local => {
+                let did_copy = if let Some((artifact_dir, done)) = artifact_dir {
+                    let se = ctx.emit_rust_step("copy JUnit test results to artifact dir", |ctx| {
+                        done.claim(ctx);
+                        let artifact_dir = artifact_dir.claim(ctx);
+                        let xmls = xmls
+                            .iter()
+                            .map(
+                                |TestResult {
+                                     junit_xml,
+                                     label,
+                                     attachments,
+                                     done: _,
+                                 }| {
+                                    (
+                                        junit_xml.clone().claim(ctx),
+                                        label.clone(),
+                                        attachments.as_ref().map(|x| {
+                                            x.iter()
+                                                .map(|(y, z)| (y.clone(), z.clone().claim(ctx)))
+                                                .collect::<BTreeMap<_, _>>()
+                                        }),
+                                    )
+                                },
+                            )
+                            .collect::<Vec<_>>();
+                        |rt| {
+                            let artifact_dir = rt.read(artifact_dir);
+
+                            for (idx, (path, label, attachments)) in xmls.into_iter().enumerate() {
+                                let Some(path) = rt.read(path) else {
+                                    continue;
```

Comment:
as written, its illegal to invoke this node with a missing junit XML, but with attachments. is that expected? if so, that constraint should be documented, and/or modeled via a tri-state enum.

(but IMO, you should just tweak this code to allow uploading attachements, regardless if there is a junit XML present)